database/bfgd/postgres/postgres.go

```diff
@@ -696,100 +614,37 @@ func (p *pgdb) L2BTCFinalityMostRecent(ctx context.Context, limit uint32) ([]bfg
 		)
 	}
 
-	tip, err := p.canonicalChainTipL2BlockNumber(ctx)
+	l2Keystones, err := p.L2KeystonesMostRecentN(ctx, limit, 0)
 	if err != nil {
 		return nil, err
 	}
 
-	// we found no canonical tip, return nothing
-	if tip == nil {
-		return []bfgd.L2BTCFinality{}, nil
-	}
-
 	finalities := []bfgd.L2BTCFinality{}
 
-	// first, get all of the most recent published finalities up to the limit
-	// from the tip
-	publishedFinalities, err := p.nextL2BTCFinalitiesPublished(
-		ctx,
-		*tip,
-		int(limit),
-	)
-	if err != nil {
-		return nil, err
-	}
-	pfi := 0
-
-	// it is possible that there will be some unpublished finalities between
-	// the published
-	// ones, get all finalities up to the limit that are NOT in published.
-	// IMPORTANT NOTE: we call these explicity "assumed unpublished"
-	// instead of explicity looking for unpublished
-	// finalities, because a finality could get published between these two
-	// queries.  this is why we call these "assumed".  the idea is to make
-	// this worst-case scenario slighty out-of-date, rather than incorrect
-	excludeL2BlockNumbers := []uint32{}
-	for _, v := range publishedFinalities {
-		excludeL2BlockNumbers = append(
-			excludeL2BlockNumbers,
-			v.L2Keystone.L2BlockNumber,
-		)
-	}
-
-	unpublishedFinalities, err := p.nextL2BTCFinalitiesAssumedUnpublished(
-		ctx,
-		*tip,
-		int(limit),
-		excludeL2BlockNumbers)
-	if err != nil {
-		return nil, err
+	hashes := []database.ByteArray{}
+	for _, l := range l2Keystones {
```

Comment:
Combine this into one query. The join should be much simpler and efficient with the table change.